src/app/components/subjectHomepage/subject-sidebar.tsx

```diff
@@ -0,0 +1,299 @@
+import { type Subject, type Unit } from "@/types";
+import { cn } from "@/lib/utils";
+import Link from "next/link";
+
+type Props = {
+  subject: Subject;
+};
+
+const SubjectSidebar = (props: Props) => {
+  const [isCollapsed, setIsCollapsed] = useState(false);
+  const [units, setUnits] = useState(props.subject.units);
+  const [editingUnit, setEditingUnit] = useState<number | null>(null);
+  const [editingChapter, setEditingChapter] = useState<{
+    unitIndex: number | null;
+    chapterIndex: number | null;
+  }>({ unitIndex: null, chapterIndex: null });
+  let sliceNum = 3;
+  const [newUnitTitle, setNewUnitTitle] = useState<string>("");
+
+  // UseRef for the input to auto-focus when editing
+  const chapterInputRef = useRef<HTMLInputElement | null>(null);
+  const unitInputRef = useRef<HTMLInputElement | null>(null);
+
+  // Auto-focus the input when editingChapter changes
+  useEffect(() => {
+    if (chapterInputRef.current) {
+      chapterInputRef.current.focus();
+    }
+
+    if (unitInputRef.current) {
+      unitInputRef.current.focus();
+    }
+  }, [editingChapter, editingUnit]);
+
+  // Handle adding a new chapter
+  const addChapter = (unitIndex: number) => {
+    if (units[unitIndex] === undefined) {
+      return;
+    }
+    const newChapter = {
+      chapter: units[unitIndex].chapters.length + 1,
+      title: "",
+    };
+    const updatedUnits = [...units];
+    updatedUnits[unitIndex]?.chapters.push(newChapter);
+    setUnits(updatedUnits);
+  };
+
+  // Handle editing a chapter's title
+  const editChapterTitle = (
+    unitIndex: number,
+    chapterIndex: number,
+    newTitle: string,
+  ) => {
+    const updatedUnits = [...units];
+    updatedUnits[unitIndex]!.chapters[chapterIndex]!.title = newTitle;
+    setUnits(updatedUnits);
+    setEditingChapter({ unitIndex: null, chapterIndex: null });
+  };
+
+  // Handle chapter delete
+  const deleteChapter = (unitIndex: number, chapterIndex: number) => {
+    const updatedUnits = [...units];
+    updatedUnits[unitIndex]?.chapters.splice(chapterIndex, 1);
+    setUnits(updatedUnits);
+  };
+
+  // Handle adding a new unit
+  const addUnit = () => {
+    const newUnit = {
+      unit: units.length + 1,
+      title: newUnitTitle || `Unit ${units.length + 1}`,
+      chapters: [],
+    };
+    setUnits([...units, newUnit]);
+    setNewUnitTitle(""); // Reset new unit title
+  };
+
+  // Handle editing a unit's title
+  const editUnitTitle = (unitIndex: number, newTitle: string) => {
+    const updatedUnits = [...units];
+    updatedUnits[unitIndex]!.title = newTitle;
+    setUnits(updatedUnits);
+    setEditingUnit(null);
+  };
+
+  // Handle unit delete
+  const deleteUnit = (unitIndex: number) => {
+    const updatedUnits = [...units];
+    updatedUnits.splice(unitIndex, 1);
+    setUnits(updatedUnits);
+  };
+
+  const pathname = window.location.pathname;
+  const isAdmin = pathname.split("/").includes("admin") ? "/admin" : "";
+  isAdmin && sliceNum++;
```

Comment:
this probably needs to be in a useEffect

Reply:
No cuz then I need to create more "use states"